File: swift/Sources/OmniTest/Fib.swift
// The system under test: a tiny Fibonacci library.
//
// Every omni port ships this same library, with the same behaviour and the
// same error messages, and tests it with the same spec file
// (spec/fib.json).

import Foundation
import Omni

/// The Fibonacci library's own error type. A subject must never throw the
/// runner's OmniError - that type is reserved for test failures.
struct FibError: Error, CustomStringConvertible {
  let message: String

  init(_ message: String) {
    self.message = message
  }

  var description: String { return message }
}

enum Fib {
  /// Validate a Fibonacci index. The spec matches on these messages.
  static func fibindex(_ val: Json) throws -> Int {
    guard let num = val.asnum else {
      throw FibError("fib: not a number: " + stringify(val))
    }

    if num != num.rounded(.towardZero) {
      throw FibError("fib: non-integer index: " + stringify(val))
    }

    if 0 > num {
      throw FibError("fib: negative index: " + stringify(val))
    }

    return Int(num)
  }

  /// The nth Fibonacci number: fib(0)=0, fib(1)=1.
  static func fibnum(_ index: Int) -> Int {
    if 0 == index {
      return 0
    }

    var prev = 0
    var cur = 1

    for _ in 1..<max(index, 1) {
      let next = prev + cur
      prev = cur
      cur = next
    }

    return cur
  }

  static func fib(_ val: Json) throws -> Json {
    return .num(Double(fibnum(try fibindex(val))))
  }

  /// The first n Fibonacci numbers.
  static func fibseq(_ val: Json) throws -> Json {
    let count = try fibindex(val)
    var out: [Json] = []
    for index in 0..<max(count, 0) {
      out.append(.num(Double(fibnum(index))))
    }
    return .list(out)
  }

  /// The Fibonacci numbers from index start to index end, inclusive.
  static func fibrange(_ start: Json, _ end: Json) throws -> Json {
    let from = try fibindex(start)
    let to = try fibindex(end)
    var out: [Json] = []
    var index = from
    while index <= to {
      out.append(.num(Double(fibnum(index))))
      index += 1
    }
    return .list(out)
  }

  /// A description of the nth Fibonacci number. `prev` is null at index 0,
  /// which exercises the runner's null handling.
  static func fibinfo(_ val: Json) throws -> Json {
    let index = try fibindex(val)
    let num = fibnum(index)

    return Json.mapOf([
      ("n", .num(Double(index))),
      ("val", .num(Double(num))),
      ("even", .bool(0 == num % 2)),
      ("prev", 0 == index ? .null : .num(Double(fibnum(index - 1)))),
      ("label", .str("fib(" + numstr(Double(index)) + ")=" + numstr(Double(num)))),
    ])
  }
}

File: go/fib/fib.go
// The system under test: a tiny Fibonacci library.
//
// Every omni port ships this same library, with the same behaviour and the
// same error messages, and tests it with the same spec file
// (spec/fib.json).

package fib

import (
	"errors"
	"math"

	omni "github.com/voxgig/omni/go"
)

// FibIndex validates a Fibonacci index. Errors are part of the contract:
// the spec matches on these messages.
func FibIndex(val any) (int, error) {
	num, isnum := omni.ToNum(val)
	if _, isbool := val.(bool); isbool || !isnum {
		return 0, errors.New("fib: not a number: " + omni.Stringify(val))
	}

	if num != math.Trunc(num) {
		return 0, errors.New("fib: non-integer index: " + omni.Stringify(val))
	}

	if num < 0 {
		return 0, errors.New("fib: negative index: " + omni.Stringify(val))
	}

	return int(num), nil
}

// Fib is the nth Fibonacci number: fib(0)=0, fib(1)=1.
func Fib(n any) (int, error) {
	index, err := FibIndex(n)
	if nil != err {
		return 0, err
	}

	if 0 == index {
		return 0, nil
	}

	prev := 0
	cur := 1

	for i := 1; i < index; i++ {
		prev, cur = cur, prev+cur
	}

	return cur, nil
}

// FibSeq is the first n Fibonacci numbers.
func FibSeq(n any) ([]any, error) {
	count, err := FibIndex(n)
	if nil != err {
		return nil, err
	}

	out := []any{}
	for i := 0; i < count; i++ {
		val, err := Fib(i)
		if nil != err {
			return nil, err
		}
		out = append(out, val)
	}

	return out, nil
}

// FibRange is the Fibonacci numbers from index start to index end,
// inclusive.
func FibRange(start any, end any) ([]any, error) {
	from, err := FibIndex(start)
	if nil != err {
		return nil, err
	}

	to, err := FibIndex(end)
	if nil != err {
		return nil, err
	}

	out := []any{}
	for i := from; i <= to; i++ {
		val, err := Fib(i)
		if nil != err {
			return nil, err
		}
		out = append(out, val)
	}

	return out, nil
}

// FibInfo describes the nth Fibonacci number. `prev` is nil at index 0,
// which exercises the runner's null handling.
func FibInfo(n any) (map[string]any, error) {
	index, err := FibIndex(n)
	if nil != err {
		return nil, err
	}

	val, err := Fib(index)
	if nil != err {
		return nil, err
	}

	var prev any
	if 0 != index {
		prevval, err := Fib(index - 1)
		if nil != err {
			return nil, err
		}
		prev = prevval
	}

	return map[string]any{
		"n":     index,
		"val":   val,
		"even":  0 == val%2,
		"prev":  prev,
		"label": "fib(" + omni.Stringify(index) + ")=" + omni.Stringify(val),
	}, nil
}
